zenoh/src/api/builders/close.rs

```diff
@@ -0,0 +1,182 @@
+//
+// Copyright (c) 2024 ZettaScale Technology
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License 2.0 which is available at
+// http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+//
+// Contributors:
+//   ZettaScale Zenoh Team, <zenoh@zettascale.tech>
+//
+
+use std::{
+    future::{Future, IntoFuture},
+    pin::Pin,
+    time::{Duration, Instant},
+};
+
+use async_trait::async_trait;
+use tokio::{select, time::sleep};
+use tracing::warn;
+use zenoh_core::{Resolvable, Wait};
+use zenoh_result::ZResult;
+use zenoh_runtime::ZRuntime;
+use zenoh_task::TaskController;
+
+/// A behavior selector for close operations used if initial timeout was reached
+pub enum CloseBackoff {
+    /// fail immediately
+    Fail,
+    /// wait for an additional timeout, then fail if timeout reached
+    Wait(Duration),
+}
+
+/// A builder for close operations.
+// NOTE: `Closeable` is only pub(crate) because it is zenoh-internal trait, so we don't
+// care about the `private_bounds` lint in this particular case.
+#[allow(private_bounds)]
+pub struct CloseBuilder<TCloseable: Closeable> {
+    closee: TCloseable::TClosee,
+    timeout: Duration,
+    backoff: CloseBackoff,
+}
+
+// NOTE: `Closeable` is only pub(crate) because it is zenoh-internal trait, so we don't
+// care about the `private_bounds` lint in this particular case.
+#[allow(private_bounds)]
+impl<TCloseable: Closeable> CloseBuilder<TCloseable> {
+    pub(crate) fn new(closeable: &TCloseable) -> Self {
+        Self {
+            closee: closeable.get_closee(),
+            timeout: Duration::from_secs(3600),
+            backoff: CloseBackoff::Fail,
+        }
+    }
+
+    /// Set the timeout for close operation
+    ///
+    /// # Arguments
+    ///
+    /// * `timeout` - The timeout value for close operation
+    ///
+    pub fn timeout(mut self, timeout: Duration) -> Self {
+        self.timeout = timeout;
+        self
+    }
+
+    /// Set the backoff for close operation
+    ///
+    /// # Arguments
+    ///
+    /// * `backoff` - The backoff behavior for close operation
+    ///
+    pub fn backoff(mut self, backoff: CloseBackoff) -> Self {
```

Comment:
What's the purpose of a `backoff` in a close operation? A backoff is usually employed as a mechanism to minimise the probability of collisions/contention.

Reply:
please check the comment above

Reply:
https://github.com/eclipse-zenoh/zenoh/pull/1576/files#r1830677025